2D3D_UnityProject/Assets/Scripts/Puzzles/Zodiac/ZodiacPuzzlePiece.cs

```diff
@@ -0,0 +1,96 @@
+    /// <summary>
+    /// Local position of piece by default and after a PieceIn()
+    /// </summary>
+    private Vector3 inPosition;
+    /// <summary>
+    /// Local position of piece after a PieceOut()
+    /// </summary>
+    private Vector3 outPosition;
+
+    private void Awake()
+    {
+        //Calculate in and out positions
+        inPosition = transform.localPosition;
+        outPosition = inPosition;
+        outPosition.y += inOutDistance;
+    }
+
+    /// <summary>
+    /// Animates a piece into the outPosition. Stops and PieceIn() that might be running.
+    /// </summary>
+    public void PieceOut()
+    {
+        if (pieceInRoutineInstance != null)
+            StopCoroutine(pieceInRoutineInstance);
+        pieceOutRoutineInstance = PieceOutRoutine();
+        StartCoroutine(pieceOutRoutineInstance);
+    }
+
+    private IEnumerator pieceOutRoutineInstance;
+    private IEnumerator PieceOutRoutine()
+    {
+        // Make an animation curve starting from current position and ending at the outPosition y values
+        AnimationCurve animationCurve = AnimationCurve.EaseInOut(0, transform.localPosition.y, inOutTime, outPosition.y);
+
+        for (float timer = 0f; timer < inOutTime; timer += Time.deltaTime)
+        {
+            // Evaluate the animation curve at the current timer value and set the current local position y to that value
+            float animValue = animationCurve.Evaluate(timer);
+            Vector3 newPosition = transform.localPosition;
+            newPosition.y = animValue;
+            transform.localPosition = newPosition;
+            yield return null;
+        }
+
+        // Jump to outPosition
+        transform.localPosition = outPosition;
+    }
+
+    /// <summary>
+    /// Animates a piece into the inPosition. Stops and PieceOut() that might be running.
+    /// </summary>
+    public void PieceIn()
+    {
+        if (pieceOutRoutineInstance != null)
+            StopCoroutine(pieceOutRoutineInstance);
+        pieceInRoutineInstance = PieceInCoroutine();
+        StartCoroutine(pieceInRoutineInstance);
+    }
+
+    private IEnumerator pieceInRoutineInstance;
+    private IEnumerator PieceInCoroutine()
+    {
+        // Make an animation curve starting from current position and ending at the outPosition y values
+        AnimationCurve animationCurve = AnimationCurve.EaseInOut(0, transform.localPosition.y, inOutTime, inPosition.y);
+
+        for (float timer = 0f; timer < inOutTime; timer += Time.deltaTime)
+        {
+            // Evaluate the animation curve at the current timer value and set the current local position y to that value
+            float animValue = animationCurve.Evaluate(timer);
+            Vector3 newPosition = transform.localPosition;
+            newPosition.y = animValue;
+            transform.localPosition = newPosition;
+            yield return null;
+        }
+
+        // Jump to inPosition
+        transform.localPosition = inPosition;
+    }
+}
```

Comment:
consolidate PieceInCoroutine() and PieceOutCoroutine() into one coroutine that accepts the target position vector (either inPosition or outPosition) as a parameter, call it something like PieceInOutCoroutine() (or whatever makes more sense than that)

then pass it the appropriate vector in the PieceIn() and PieceOut() methods, that way we have the movement logic in one spot

![image](https://user-images.githubusercontent.com/23004127/68537109-da0dc000-032c-11ea-9614-7f2b7f89e643.png)
something like this